File: kmp/composeApp/src/commonMain/kotlin/com/tanyudii/tmuxweb/domain/SessionFilter.kt
package com.tanyudii.tmuxweb.domain

import com.tanyudii.tmuxweb.domain.model.ProjectSession

/**
 * EMB-221: session-list status filter. "Status" here means whether the
 * session is currently attached (someone has a terminal open on it) --
 * the only real-time status signal already present on every
 * [ProjectSession] with no extra per-session network round trip. Live
 * environment (docker) running/idle status needs a `docker compose ps`
 * call per session (see SessionResourceUsageViewModel) and isn't cheap
 * enough to run for every row in a list, so it's out of scope here.
 */
enum class SessionStatusFilter { ALL, ACTIVE, IDLE }

/**
 * Pure, DOM-free filter so it's directly testable from commonTest -- same
 * split as [isFindShortcut]/[isCopyShortcut]. Branch matching is a
 * case-insensitive substring match against [ProjectSession.name], which
 * doubles as the git branch name for every session (see
 * `slugifyBranchName`/`addWorktree` in src/project-sessions.ts) -- there's
 * no separate `branch` field on [ProjectSession] to filter on instead.
 */
fun filterSessions(
    sessions: List<ProjectSession>,
    statusFilter: SessionStatusFilter,
    branchQuery: String,
): List<ProjectSession> {
    val query = branchQuery.trim()
    return sessions.filter { session ->
        val matchesStatus = when (statusFilter) {
            SessionStatusFilter.ALL -> true
            SessionStatusFilter.ACTIVE -> session.attached
            SessionStatusFilter.IDLE -> !session.attached
        }
        val matchesBranch = query.isEmpty() || session.name.contains(query, ignoreCase = true)
        matchesStatus && matchesBranch
    }
}

File: web/src/domain/sessionFilter.ts
// Ports kmp/composeApp/.../domain/SessionFilter.kt (EMB-221): session-list
// status filter. "Status" here means whether the session is currently
// attached -- the only real-time status signal already present on every
// ProjectSession with no extra per-session network round trip.
import type { ProjectSession } from "../api/types";

export type SessionStatusFilter = "all" | "active" | "idle";

/**
 * Pure, DOM-free filter. Branch matching is a case-insensitive substring
 * match against `session.name`, which doubles as the git branch name for
 * every session (see `slugifyBranchName`/`addWorktree` in
 * src/project-sessions.ts) -- there's no separate `branch` field to filter
 * on instead.
 */
export function filterSessions(
  sessions: ProjectSession[],
  statusFilter: SessionStatusFilter,
  branchQuery: string,
): ProjectSession[] {
  const query = branchQuery.trim().toLowerCase();
  return sessions.filter((session) => {
    const matchesStatus =
      statusFilter === "all" ? true : statusFilter === "active" ? session.attached : !session.attached;
    const matchesBranch = query.length === 0 || session.name.toLowerCase().includes(query);
    return matchesStatus && matchesBranch;
  });
}
